### ISSUE TITLE
TLS1.3 ClientHello.legacy_version is not verified

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#6600]
from [draft-28](https://tools.ietf.org/html/draft-ietf-tls-tls13-28#page-139):
```
   Implementations MUST NOT send a ClientHello.legacy_version or
   ServerHello.legacy_version set to 0x0300 or less.  Any endpoint
   receiving a Hello message with ClientHello.legacy_version or
   ServerHello.legacy_version set to 0x0300 MUST abort the handshake
   with a "protocol_version" alert.
```
current master (358ffa05cd3a0888) ignores ClientHello.legacy_version and continues the connection.

  <details><summary>Reproducer</summary><p>

```
openssl req -x509 -newkey rsa -keyout localhost.key -out localhost.crt -subj /CN=localhost -nodes -batch
openssl s_server -key localhost.key -cert localhost.crt -www 2>server.err >server.out &
openssl_pid=$!
git clone https://github.com/tomato42/tlsfuzzer
pushd tlsfuzzer
git clone https://github.com/tomato42/tlslite-ng .tlslite-ng
ln -s .tlslite-ng/tlslite tlslite
git clone https://github.com/warner/python-ecdsa .python-ecdsa
ln -s .python-ecdsa/ecdsa ecdsa
PYTHONPATH=. python scripts/test-tls13-legacy-version.py
popd
kill $openssl_pid
```
  </p></details>

  <details><summary>tlsfuzzer output</summary><p>

```
sanity ...
OK

version (1, 0) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c329d0> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c32550>) with last message being: <tlslite.messages.Message object at 0x7f4b42bb9e10>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (0, 2) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c32290> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c32610>) with last message being: <tlslite.messages.Message object at 0x7f4b42bde9d0>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (3, 0) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c1fe10> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c1ff50>) with last message being: <tlslite.messages.Message object at 0x7f4b42bb9410>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (0, 0) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c32190> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c320d0>) with last message being: <tlslite.messages.Message object at 0x7f4b42bde910>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (0, 1) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42bb9090> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42bb9150>) with last message being: <tlslite.messages.Message object at 0x7f4b42bdea90>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (1, 2) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c32f10> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c32e50>) with last message being: <tlslite.messages.Message object at 0x7f4b42bde7d0>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (2, 0) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c32450> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c32390>) with last message being: <tlslite.messages.Message object at 0x7f4b42bdea90>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

version (1, 3) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f4b42c32a50> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f4b42c32a90>) with last message being: <tlslite.messages.Message object at 0x7f4b42bde710>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-legacy-version.py", line 170, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Handshake(server_hello)

sanity ...
OK

Legacy version in TLS 1.3
Check that invalid legacy versions are rejected
with protocol version alert by the TLS 1.3 server.

version: 1

Test end
successful: 2
failed: 8
  'version (0, 0)'
  'version (0, 1)'
  'version (0, 2)'
  'version (1, 0)'
  'version (1, 2)'
  'version (1, 3)'
  'version (2, 0)'
  'version (3, 0)'
```
  </p></details>

**Comment by mattcaswell** [comment]
This is deliberate because we have not removed support for SSLv3. Arguably what we should do is only support SSLv3 if the max protocol is set to TLSv1.2 or below.

**Comment by tomato42** [comment]
the test also checks for values lower than 0x0300, and they aren't rejected either

also, the ClientHello messages sent are valid TLS 1.3 ClientHello messages, that is, with supported_versions extension, so rejecting them, even if SSLv3 is still supported, should not cause interoperability issues for SSLv3 implementations...

(not that I consider this to be a pressing or important issue, I've filed it mostly for completeness)

**Comment by mattcaswell** [comment]
> the test also checks for values lower than 0x0300, and they aren't rejected either

I think we just completely ignore legacy_version if supported_versions is present. Should the test be checking for values lower than 0x0300? The spec text that you posted above only mentions 0x0300.

**Comment by mattcaswell** [comment]
The more I think about it, the more I think this is a fairly pointless requirement in the spec. Since it has absolutely no impact on interoperability with SSLv3 clients why does this matter?

**Comment by mattcaswell** [comment]
I'm thinking we should just ignore this issue. The requirement doesn't make any sense to me; shouldn't ever happen in practice; and doesn't really matter even if it did.

**Comment by kaduk** [comment]
I think the requirement is mostly about keeping the ecosystem clean; it would be nice to have some way to know that there is not "random junk" or nonsensical values being received here, and the spec's text gives us cover to abort instead of heeding Postel.

**Comment by tomato42** [comment]
> I think we just completely ignore legacy_version if supported_versions is present. Should the test be checking for values lower than 0x0300? 

The text is quite explicit that the client MUST NOT send those, so this is a protocol violation. For protocol violations we want to have a strict expected behaviour as departure from them (in future) likely means regression, and possibly a security issue.

Given that the receiving of 0x0300 is a MUST abort, I consider it form implementation point of view to be easier to have the same behaviour for all the lower values too. And more consistent from purely logical point of view.

Note: tlsfuzzer is a TLS _compliance_ test suite, I'm not writing what is the behaviour of specific implementation and then test others, I'm using the specification to check if different implementations follow it strictly.

> and doesn't really matter even if it did.

yes, _shouldn't_ matter, but making sure that some clients don't get confused and start sending 0x0103 that is then accepted by some implementations (openssl) and rejected by others (nss, gnutls) is even worse (see also: [Postel was wrong](https://tools.ietf.org/html/draft-thomson-postel-was-wrong-03))

**Comment by richsalz** [comment]
If it's a protocol violation, then the server isn't required to do anything.  It's like C's "undefined behavior," right?

Postel or XML-style....  how many angels can dance?

**Comment by mattcaswell** [comment]
I think the value of this check is somewhat debatable. However, it isn't a difficult check and does not particularly increase complexity. Therefore I added a fix in #6747.
